projects/solar/src/relay_fsm.c

```diff
@@ -2,17 +2,64 @@
 
 #include <stdbool.h>
 
+#include "can_transmit.h"
+#include "can_unpack.h"
+#include "data_store.h"
 #include "event_queue.h"
 #include "fault_handler.h"
 #include "fsm.h"
+#include "gpio.h"
 #include "log.h"
+#include "solar_boards.h"
 #include "solar_config.h"
 #include "solar_events.h"
 #include "status.h"
 
+static SolarMpptCount solar_mppt_count;
+
+static void prv_assert_relay(SoftTimerId timer_id, void *context) {
+  RelayFsmStorage *storage = (RelayFsmStorage *)context;
+  if (storage->isErrCalled == true) {
+    fault_handler_raise_fault(EE_SOLAR_RELAY_OPEN_ERROR, 0);
+    return;
+  }
+
+  int32_t data_value;
+  bool isSet = false;
+
+  data_store_get_is_set(DATA_POINT_CURRENT, &isSet);
+
+  if (!isSet) {
+    if (storage->isSetCounter >= 30) {
+      LOG_DEBUG("Aborting, The current is not set\n");
+      fault_handler_raise_fault(EE_SOLAR_RELAY_OPEN_ERROR, 0xFF);
+      return;
+    }
+    storage->isSetCounter += 1;
+    soft_timer_start_millis(DATA_STORE_ASSERTION_DELAY_MS, prv_assert_relay, context, NULL);
+    return;
+  }
+
+  data_store_get(DATA_POINT_CURRENT, (uint32_t *)&data_value);
+
+  if (data_value > CURRENT_ASSERT_THRESHOLD_uA) {
+    fault_handler_raise_fault(EE_SOLAR_RELAY_OPEN_ERROR, data_value);
```

Comment:
I'm not sure that your scheme of putting the current value in the fault data will work, because the fault data is a uint8 while the current is an int16, so it won't fit. I guess we could edit the solar fault CAN message to make the data field wider, but that could be a pain - I think it might be better to just signal that there was an overcurrent fault (and not specify what the current is).

To do this, add an enum in exported_enums.h beside the other solar ones and give it a good name with three values - one for "drv120 error", one for "current exceeded / relay not opened", and one for "current not set after timeout" (i.e. the >= 30 case above).

Reply:
could you elaborate on this? the `drv120_error` and `relay not opened` errors already exist. so would I just add two more which would be `current exceeded and `current not set after timeout` errors but should I make new errors for all with some common word like:
`RELAY_ERROR_DRV120`
`RELAY_ERROR_RELAY_NOT_OPEN`
`RELAY_ERROR_CURRENT_EXCEDDED`
`RELAY_ERROR_CURRENT_NEVER_SET`

Reply:
I meant another enum in exported_enums.h in addition to the `EESolarError` enum just for data values for `EE_SOLAR_FAULT_RELAY_OPEN_ERROR` faults. So you'd have an enum maybe called `EESolarRelayOpenErrorReason` with members like `EE_SOLAR_RELAY_ERROR_DRV120`, `EE_SOLAR_RELAY_ERROR_CURRENT_EXCEEDED_NOT_OPEN`, `EE_RELAY_ERROR_CURRENT_NEVER_SET` (and "relay not opened/current exceeded" should be one error since it's sent from the same place).

Reply:
done